mitosheet/src/components/LoadingIndicator.tsx

```diff
@@ -1,43 +1,275 @@
+            return 'Updating user profile'
+        }
+        case UpdateType.SetUserFieldUpdate: {
+            return 'Updating user profile'
+        }
+        case UpdateType.UpdateFeedbackv2ObjObject: {
+            return 'Updating user profile'
+        }
+        case UpdateType.GoPro: {
+            return 'Activating Mito Pro'
+        }
+        case UpdateType.RenderCountUpdate: {
+            return 'Updating user profile'
+        }
+    }
+}
+
+const getMessageType = (messageType: string): StepType | UpdateType | undefined => {
+    if (isStepEvent(messageType)) {
+        return messageType.substring(0, messageType.length - "_edit".length) as StepType;
+    } else if  (isUpdateEvent(messageType)) {
+        return messageType as UpdateType;
+    }
+    
+    return undefined;
+}
+
+const getMessageTypesToDisplay  = (loading: [string, string | undefined, string][]): [(StepType | UpdateType), string][] => {
+    const seenStepIds: string[] = []
+    return loading.filter((([, step_id, ]) => {
+        // We filter out any edits that have duplicated step ids, as users think of 
+        // these as just a single edit. We only take the first one!
+        if (step_id !== undefined) {
+            if (seenStepIds.includes(step_id)) {
+                return false;
+            }
+            seenStepIds.push(step_id);
+        }
+        return true;
+
+    })).map((([message_id, , type]) => {
+        return [getMessageType(type), message_id];
+    })).filter(([messageType, ]) => {
+        return messageType !== undefined;
+    }) as [(StepType | UpdateType), string][];
+}
+
+
 
 /*
-    A tiny, upper left modal that tells the user that the operation is
-    loading.
+    Gives the user lots of information about what events and updates
+    and loading.
 
     By default, does not displaying anything for the first .5 seconds it
-    is rendered, so that only long running ops actually display a loading
-    bar.
+    is rendered, so that only long running ops actually display anything.
 */
-const LoadingIndicator = (): JSX.Element => {
-    // We use a count to track the number of ...s to display.
-    // 0 -> '', 1 -> '.', 2 -> '..', 3 -> '...'. Wraps % 4.
+const LoadingIndicator = (props: {loading: [string, string | undefined, string][]}): JSX.Element => {
     const [display, setDisplay] = useState(false);
 
-    // Only display this after 500 ms
+    // We store the message at the top of the loading queue, so that we can 
+    // track if it has been running for longer than 10 seconds
+    const [currentLoadingMessage, setCurrentLoadingMessage] = useState<undefined | [number, string]>(undefined);
+
+    // This makes sure we're only displaying after .5 seconds of loading
+    useEffect(() => {
+        if (props.loading.length === 0) {
+            setDisplay(false);
+        } else if (props.loading.length > 0) {
+            const timeout = setTimeout(() => {
+                setDisplay(true);
+            }, 500);
```

Comment:
When I'm using the tool this seems long. But I suppose it isn't.

Reply:
This is fine as is! I was just paying too much attention to the loading indicator and waiting for it to appear I think.